keps/sig-node/3857-rro-hostpath/README.md

```diff
@@ -0,0 +1,967 @@
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+The current `readOnly` hostPath volumes are not recursively read-only;
+e.g., even if `/mnt` is mounted as read-only, its submounts such as `/mnt/usbstorage` are not read-only.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+Support recursively read-only mounts for runc >= 1.1 && kernel >= 5.12.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+Support recursively read-only mounts for old runc and old kernel releases.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wants to mount `/mnt`, includings its submounts such as `/mnt/usbstorage`, as read-only.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+Constraints: needs runc >= 1.1 && kernel >= 5.12.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Risk: a manifest with `recursivelyReadOnly: Required` will not work on old distros (See "Constraints" above).
+
+Mitigation: set `recursivelyReadOnly: IfPossible`.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### Core API
```

Comment:
https://github.com/kubernetes/enhancements/pull/3858/files#discussion_r1102038347

To make this KEP viable we need something like `pod.status.volumeStatuses` which indicates the result.

We MAY want something in Node (to indicate support), but that's less obvious to me - @bobbypage

Reply:
+1 on indicating the status. Ideally the user stories section needs to explain step-by-step user experience.

Reply:
The KEP is for [`VolumeMount`](https://pkg.go.dev/k8s.io/api/core/v1#VolumeMount) structs, not for [`Volume`](https://pkg.go.dev/k8s.io/api/core/v1#VolumeMount) structs. 
So probably it should be visible in [`ContainerStatus`](https://pkg.go.dev/k8s.io/api/core/v1#ContainerStatus) structs?

Reply:
Added `ContainerStatus.[]VolumeMounts.RecursivelyReadonly`
```diff
diff --git a/pkg/apis/core/types.go b/pkg/apis/core/types.go
index 04ffd250b61..3887f940ee6 100644
--- a/pkg/apis/core/types.go
+++ b/pkg/apis/core/types.go
@@ -1878,6 +1878,9 @@ type VolumeMount struct {
        // SubPathExpr and SubPath are mutually exclusive.
        // +optional
        SubPathExpr string
+       // When RecursivelyReadOnly is set to a non-nil value, ReadOnly must be set to true.
+       // +optional
+       RecursivelyReadOnly *RecursivelyReadOnlyMode
 }
 
 // MountPropagationMode describes mount propagation.
@@ -1905,6 +1908,18 @@ const (
        MountPropagationBidirectional MountPropagationMode = "Bidirectional"
 )
 
+// RecursivelyReadOnlyMode describes recursively-readonly mode.
+type RecursivelyReadOnlyMode string
+
+const (
+       // RecursivelyReadOnlyDisabled disables recursively-readonly mode.
+       RecursivelyReadOnlyDisabled RecursivelyReadOnlyMode = "Disabled"
+       // RecursivelyReadOnlyDisabled enables recursively-readonly mode if possible.
+       RecursivelyReadOnlyIfPossible RecursivelyReadOnlyMode = "IfPossible"
+       // RecursivelyReadOnlyDisabled enables recursively-readonly mode, or raise an error.
+       RecursivelyReadOnlyEnabled RecursivelyReadOnlyMode = "Enabled"
+)
+
 // VolumeDevice describes a mapping of a raw block device within a container.
 type VolumeDevice struct {
        // name must match the name of a persistentVolumeClaim in the pod
@@ -2505,6 +2520,9 @@ type ContainerStatus struct {
        // +featureGate=InPlacePodVerticalScaling
        // +optional
        Resources *ResourceRequirements
+       // Status of volume mounts.
+       // +optional
+       VolumeMounts []VolumeMountStatus
 }
 
 // PodPhase is a label for the condition of a pod at the current time.
@@ -2584,6 +2602,17 @@ const (
        PodResizeStatusInfeasible PodResizeStatus = "Infeasible"
 )
 
+// VolumeMountStatus shows status of volume mounts.
+type VolumeMountStatus struct {
+       // Corresponds to the name of the original VolumeMount.
+       Name string
+       // RecursivelyReadOnly must be RecursivelyReadOnlyDisabled, RecursivelyReadOnlyEnabled, or nil.
+       // RecursivelyReadOnlyIfPossible in the original VolumeMount must be translated to
+       // RecursivelyReadOnlyDisabled or RecursivelyReadOnlyEnabled, depending on the mount result.
+       // +optional
+       RecursivelyReadOnly *RecursivelyReadOnlyMode
+}
+
 // RestartPolicy describes how the container should be restarted.
 // Only one of the following restart policies may be specified.
 // If none of the following policies is specified, the default one
```

Reply:
@thockin @SergeyKanzhelev LGTY?

Reply:
thinking more about it - status is mostly needed because there is no guarantee that it will be applied. So if there will be no `IfPossible`, than no status will be needed.

Reply:
It is likely that we have to keep `IfPossible` until 2029 or maybe 2030
https://github.com/kubernetes/enhancements/pull/3858#discussion_r1195408670

Reply:
The staus needs to be keyed on the mountPath - the same volume can be mounted multiple times on different paths with different modes.  It's bizarre, but it works.  You probably want to keep both volument name and mount path.

Reply:
Changed from `[]VolumeMountStatus` to `map[string]VolumeMountStatus`, and added `MountPath` to `type VolumeMountStatus struct{}`